### PR TITLE
fix logical error in statistics estimation

### LINKED ISSUE #67742: Logical error: 'Statistics 'tdigest' does not support estimating value of type String'.
https://s3.amazonaws.com/clickhouse-test-reports/67727/fc651cc0c61feb37e9cf104612cc0ac0cd7448e9/ast_fuzzer__ubsan_.html

Revert this: https://github.com/ClickHouse/ClickHouse/pull/59357

### PR DESCRIPTION
**Pull request by hanfei1991**
fix https://github.com/ClickHouse/ClickHouse/issues/67742

This PR converts type and deal with type incompatible in estimator 

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit bc2134d9df6e572009bdc6cab5072a8331dfa600 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67858/bc2134d9df6e572009bdc6cab5072a8331dfa600/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/67858/bc2134d9df6e572009bdc6cab5072a8331dfa600/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

File: src/Storages/Statistics/Statistics.h [hunk 1]
```diff
@@ -46,6 +45,8 @@ class IStatistics
     virtual Float64 estimateEqual(const Field & val) const; /// cardinality of val in the column
     virtual Float64 estimateLess(const Field & val) const;  /// summarized cardinality of values < val in the column
 
+    virtual DataTypePtr getTargetType() const { return nullptr; }
```
Comment (rschu1ze) [inline src/Storages/Statistics/Statistics.h:48]: To the reader, it is not clear what the "target type", respectively `getTargetType`, represents. Please add a comment. The comment ideally also describes what `nullptr` means as a target type.

File: src/Storages/Statistics/ConditionSelectivityEstimator.cpp [hunk 2]
```diff
@@ -14,6 +15,34 @@ void ConditionSelectivityEstimator::ColumnSelectivityEstimator::merge(String par
     if (part_statistics.contains(part_name))
         throw Exception(ErrorCodes::LOGICAL_ERROR, "part {} has been added in column {}", part_name, stats->columnName());
     part_statistics[part_name] = stats;
+    stats->setTargetType(statistics_target_types);
+}
+
+bool ConditionSelectivityEstimator::ColumnSelectivityEstimator::convertToTargetType(Field & val, const String & op) const
+{
+    auto tdigest_type_it = statistics_target_types.find(StatisticsType::TDigest);
```
Comment (rschu1ze) [inline src/Storages/Statistics/ConditionSelectivityEstimator.cpp:23]: The implementation of `convertToTargetType` makes some assumptions how which statistics types are able to estimate the operator `op`. The logic is similar to `ColumnStatistics::estimateLess` and `ColumnStatistics::estimateEqual`. This creates an implicit dependency and duplication where we ideally have this code in one single code.

**Comment by rschu1ze** [comment]
To be honest, the other PR #67013 seems to solve #67742 in a more straightforward way (at least for me). The other PR "only" introduces a new statistics type \+ does some refactoring, it is almost ready - maybe we should merge it instead.

**Pull request closed without merging**